### PR TITLE
feat: db upgrade

### LINKED ISSUE #6: Choosing the DB and building abstractions for it
At the minute, we use nedb as our database. This package is outdated and unmaintained so we need to move to a new one.
This DB does not (and for the sake of complexity, will not) be hosted as we do not need such high throughput or feature sets.

When we choose the DB, we should make stable abstractions around it so that the backend data structure can be swapped at any time without breaking the bot. As it stands, such a move would be impossible.

- [ ] Wait for <https://github.com/TheCodingDen/firebase-killswitch> to be developed
- [ ] Integrate Firebase SDK
- [ ] Build database abstractions

### PR DESCRIPTION
**Pull request by nullishamy**
This upgrades the DB by changing to Postgres, implementing Prisma, and refactoring the DB implementation to comply with the above.

This will be used with Heroku's DB implementation to provide easy persistence whilst the project is hosted in production.

Closes #6

### CODE REVIEW

File: src/managers/submissions.ts [hunk 1]
```diff
@@ -29,121 +29,215 @@ export class SubmissionsManager extends Manager {
...
     }
 
-    log.debug(`Loaded project from DB: ${JSON.stringify(value)}`)
+    log.debug(`Loaded project from DB: ${JSON.stringify(data)}`)
+
+    // TS cant infer that details is not null here
+    // Thus, we cast to this type and tell it that
+    // Though, if other type errors remain, they will throw errors
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+    const detailsNotNull = data as typeof data & { details: SubmissionDetails }
```
Comment (MaxInertia) [inline src/managers/submissions.ts:111]: We could use a type guard instead. That way the check lives alongside the type refinement. and the intent is clear without requiring these comments.
```ts
function hasDetails<D, T extends {details: D | null}>(x: T): x is T & {details: NonNullable<T['details']>} {
  return x.details !== null
}
```
Then we replace this
```ts
if(!data.details) {
```
with this
```ts
if(!hasDetails(data)) {
```

and remove each instance of these five lines

File: src/managers/submissions.ts [hunk 2]
```diff
@@ -29,121 +29,215 @@ export class SubmissionsManager extends Manager {
...
+        log.error(msg)
+        throw new Error(msg)
+      }
 
-    return await Promise.all(data.map(async d => await this.resolve(d)))
+      // TS cant infer that details is not null here
+      // Thus, we cast to this type and tell it that
+      // Though, if other type errors remain, they will throw errors
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+      const detailsNotNull = d as typeof d & { details: SubmissionDetails }
```
Comment (MaxInertia) [inline src/managers/submissions.ts:207]: ditto

File: src/managers/submissions.ts [hunk 3]
```diff
@@ -29,121 +29,215 @@ export class SubmissionsManager extends Manager {
...
+      log.error(msg)
+      throw new Error(msg)
     }
 
-    return await this.resolve(value)
+    // TS cant infer that details is not null here
+    // Thus, we cast to this type and tell it that
+    // Though, if other type errors remain, they will throw errors
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+    const detailsNotNull = data as typeof data & { details: SubmissionDetails }
```
Comment (MaxInertia) [inline src/managers/submissions.ts:239]: ditto

**Review by MaxInertia — COMMENTED** [review]
Suggestion to use custom type guard instead of type assertion

File: src/managers/submissions.ts (hunk 1 above)
Comment (nullishamy) [reply src/managers/submissions.ts]: Thanks, that seems to work. Would you happen to know why it was unable to infer originally?

File: src/models/submission.ts [hunk 4]
```diff
@@ -112,34 +127,65 @@ export class Submission extends Model<StoredSubmissionData> {
...
   }
 
-  toSerialised (): StoredSubmissionData {
+  private flattenVotes (): PrismaVote[] {
+    return [
+      ...this.votes.upvotes.staff.map(v => v.toSerialised()),
+      ...this.votes.upvotes.veterans.map(v => v.toSerialised()),
+      ...this.votes.downvotes.staff.map(v => v.toSerialised()),
+      ...this.votes.downvotes.veterans.map(v => v.toSerialised())
+    ]
```
Comment (BobobUnicorn) [inline src/models/submission.ts:172]:
```
type Cross<T extends readonly unknown[], U extends readonly unknown[]> = Flatten<{
  [K in keyof T]: {
    [J in keyof U]: [T[K], U[J]]
  }
}>

type Flatten<T extends readonly unknown[], Acc extends readonly unknown[] = readonly []> = 
  T extends readonly [infer H, ...infer T] ?
    H extends readonly unknown[] ?
      Flatten<T, readonly [...Acc, ...H]> :
      Flatten<T, readonly [...Acc, H]> :
    Acc;

function crossProduct<T extends readonly unknown[], U extends readonly unknown[]>(a: readonly [...T], z: readonly [...U]): Cross<T, U> {
  return a.flatMap(aa => z.map(zz => [aa, zz])) as unknown as Cross<T, U>;
}

return crossProduct(['upvotes', 'downvotes'] as const, ['staff', 'veterans'] as const).flatMap(([a, b]) => votes[a][b]);
```

**Review by BobobUnicorn — CHANGES_REQUESTED** [review]

File: src/managers/submissions.ts (hunk 1 above)
Comment (MaxInertia) [reply src/managers/submissions.ts]: Nope. I was expecting the type to be refined just like in this case: [TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvARgQwE4C4oDOw2AlgHYDmUAPlGQK4A2jAvgLABQnAZvWQMbAScMlHoEI8OAApuCfFICU+IqUppOUKCW7SAhHLgA6LNkUaOWrdgjB62UQCJHmqOy6WoNuw6iGTOADcrgD0IVYRVgB6APyc7jx8gsKi4pIIAOokwAAWACo52HAA7rLysAjKhMTkVKiuOvr+pub1nlq5RcV0EN0AothF2NKOxZgA1uOYjoquCZ7e9qLNQaHhkRGx8ZxAA)

Do you know why @BobobUnicorn?

Comment (BobobUnicorn) [reply src/managers/submissions.ts]: I explained it in voice chat, but essentially TypeScript won't propagate the implicit type guard/type narrowing from the type of the inner property to the outer type; `foo.bar` is no longer `null` but `foo` will still be `Foo`

Comment (MaxInertia) [reply src/managers/submissions.ts]: For some reason I thought this was added in a recent TS version, but looking through them I don't see it.

**Review by BobobUnicorn — APPROVED** [review]

**Pull request merged**